CondTools/SiStrip/plugins/SiStripApvGainFromFileBuilder.cc

```diff
@@ -0,0 +1,637 @@
+        LogTrace("SiStripApvGainFromFileBuilder") << "output gain:" << summary.gain_in_db << std::endl;
+      } catch (std::exception& e) {
+        std::cerr << e.what() << std::endl;
+        edm::LogError("MappingError") << "@SUB=analyze"
+                                      << "Job end prematurely." << std::endl;
+        return;
+      }
+    }
+
+    SiStripApvGain::Range range(theSiStripVector.begin(), theSiStripVector.end());
+    if (!obj->put(it.first, range))
+      edm::LogError("IndexError") << "@SUB=analyze"
+                                  << "detid already exists." << std::endl;
+  }
+
+  if (outputSummary_)
+    output_summary();
+
+  //End now write sistripnoises data in DB
+  edm::Service<cond::service::PoolDBOutputService> mydbservice;
+
+  if (mydbservice.isAvailable()) {
+    if (mydbservice->isNewTagRequest("SiStripApvGainRcd")) {
+      mydbservice->createOneIOV<SiStripApvGain>(*obj, mydbservice->beginOfTime(), "SiStripApvGainRcd");
+    } else {
+      mydbservice->appendOneIOV<SiStripApvGain>(*obj, mydbservice->currentTime(), "SiStripApvGainRcd");
+    }
+  } else {
+    edm::LogError("DBServiceNotAvailable") << "@SUB=analyze"
+                                           << "DB Service is unavailable" << std::endl;
+  }
+}
+
+void SiStripApvGainFromFileBuilder::read_tickmark() {
+  // Connect file for input
+  const auto& filename = tfp_.fullPath();
+  std::ifstream thickmark_heights(filename.c_str());
+
+  if (!thickmark_heights.is_open()) {
+    edm::LogError("FileNotFound") << "@SUB=read_ticlmark"
+                                  << "File with thickmark height file " << filename.c_str() << " cannot be opened!"
+                                  << std::endl;
+    return;
+  }
+
+  // clear internal maps
+  for_each(gains_.begin(), gains_.end(), CleanUp);
+  for_each(negative_gains_.begin(), negative_gains_.end(), CleanUp);
+  for_each(null_gains_.begin(), null_gains_.end(), CleanUp);
+
+  // read file and fill internal map
+  uint32_t det_id = 0;
+  uint32_t APV_id = 0;
+  float tick_h = 0.;
+
+  int count = -1;
+
+  for (;;) {
+    count++;
+    thickmark_heights >> det_id >> APV_id >> tick_h;
+
+    if (!(thickmark_heights.eof() || thickmark_heights.fail())) {
+      if (count == 0) {
+        LogTrace("Debug") << "Reading " << filename.c_str() << " for gathering the tickmark heights" << std::endl;
+        LogTrace("Debug") << "|  Det Id   |  APV Id  |  Tickmark" << std::endl;
+        LogTrace("Debug") << "+-----------+----------+----------" << std::endl;
+      }
+      LogTrace("Debug") << std::setw(11) << det_id << std::setw(8) << APV_id << std::setw(14) << tick_h << std::endl;
+
+      // retrieve the map corresponding to the gain collection
+      Gain* map = nullptr;
+      if (tick_h > 0.) {
+        map = get_map(&gains_, APV_id);
+      } else if (tick_h < 0.) {
+        map = get_map(&negative_gains_, APV_id);
+      } else if (tick_h == 0.) {
```

Comment:
```suggestion
      } else  {
```
I am often suspicious about equality checks between non-integer numbers: did you verify whether this case really intercepts all "0"'s listed in the data file?

Here `thick_h` needs to be converted from `float` to `double` in order to be compared with "0.": if the above check did not succeed, even in a few cases, could you try replacing the comparisons above with `> 0.f` and `< 0.f` ?

Reply:
Hi @perrotta I've taken your comments and implemented them, though I have checked that with or without them I obtain bitwise the same payloads and summary output ASCII files.

Reply:
Thank you @mmusich for the check: did you also verify if all (or at least any) of the 0's is correctly propagated in the maps?
By the way, since you updated all other instances, you could do the same here, removing the (supposedly) useless `if` , or requiring it being equal to `0.f`

Reply:
>  did you also verify if all (or at least any) of the 0's is correctly propagated in the maps?

yes.

>  By the way, since you updated all other instances, you could do the same here, removing the (supposedly) useless `if` , or requiring it being equal to `0.f`

done in the last push.